[TVMScript] `T.match_buffer` syntax sugar in arguments for TVMScript printer

cyx-6 opened the pull request:
This PR implements the syntax sugar of `T.match_buffer` for new TVMScript printer. This syntax sugar will replace the `T.handle` in `T.prim_func` arguments, with matched simple buffer. For example, it will change
```python
@T.prim_func
def func(a: T.handle, b: T.handle, c: T.handle):
  A = T.match_buffer(a, [128], dtype="float32")
  B = T.match_buffer(b, [128, 128], dtype="int32")
  C = T.match_buffer(c, [128, 128, 128], dtype="uint8")
```
into
```python
@T.prim_func
def main(A: T.Buffer[(128,)], B: T.Buffer[(128, 128), "int32"], C: T.Buffer[(128, 128, 128), "uint8"]):
  T.evaluate(0)
```

junrushao commented on src/script/printer/tir/buffer.cc:
```diff
@@ -126,6 +126,18 @@ ExprDoc BufferDecl(const tir::Buffer& buffer, const String& method, const Array<
+      indices_doc.push_back(doc.value());
+    }
+  }
+  return TIR("Buffer")[indices_doc];
```
Let's modernize it by using `->Call` instead!

cyx-6 replied:
but `->Call` may generate `T.Buffer()` instead of `T.Buffer[]`?

junrushao replied:
Yep, and it's more recommended that `T.Buffer[]`

junrushao commented:
You might want to quickly run a small subset of unittests to confirm if it works by:

```python
pytest -vv tests/python/unittest/test_tvmscript_*
```

Currently the followings are failing:


```
FAILED tests/python/unittest/test_tvmscript_complete.py::test_complete_buffer_indices - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_complete.py::test_complete_alloc_buffer - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_ops.py::test_alloc_zero_dim_buffer_round_trip - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_printer_tir.py::test_prim_func - AssertionError: assert ('@T.prim_func\n'\n 'def main(A: T.Buffer[(128, 128)], B: T.Buffer[(256, 256)]):\n'\n '    T.evaluate(0)') == ('@T.prim_func\n'\n 'def ...
FAILED tests/python/unittest/test_tvmscript_regression.py::test_multi_element_array_in_outmost_namespace - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_matmul_original - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_element_wise - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_predicate - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_for_thread_binding - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_match_buffer_region - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_block_elements - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_opaque_block - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_same_name_var - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[opt_gemm_normalize] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[opt_gemm_lower] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[opt_conv_tensorcore_normalize] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[opt_conv_tensorcore_lower] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[vthread_func] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[matmul] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[rank0] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[rank0_block] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[select] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[minmax] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[abs] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[constant_folding] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[while_loop] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[comm_reducer_single_reduce_group] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[comm_reducer_multiple_reduce_groups] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[nontrivial_range_axis] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[int64_support] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[buffer_axis_separator] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[buffer_ramp_access_as_slice_index] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[decl_buffer] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[allocate_and_decl_buffer] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[float_infinity] - tvm.error.DiagnosticError: Traceback (most recent call last):
FAILED tests/python/unittest/test_tvmscript_roundtrip.py::test_roundtrip[multi_env_threads] - tvm.error.DiagnosticError: Traceback (most recent call last):
```

cyx-6 committed:
remap syntax sugar for tvmscript printer

fix outdated code

fix outdated unittest

`match_buffer` syntax sugar

fix

fix bugs

cyx-6 commented on src/script/printer/tir/buffer.cc:
```diff
@@ -126,6 +126,14 @@ ExprDoc BufferDecl(const tir::Buffer& buffer, const String& method, const Array<
                     /*args=*/args);
 }
 
+ExprDoc BufferAttn(const tir::Buffer& buffer, const ObjectPath& p, const Frame& frame,
```
annotation, like
```python
A : T.Buffer() = ...
```

junrushao replied:
Oh that makes sense. Thanks!

junrushao commented on src/script/printer/tir/function.cc:
```diff
@@ -34,16 +36,54 @@ String FindFunctionName(const IRDocsifier& d, const tir::PrimFunc& f) {
       args.reserve(n_args);
+      std::unordered_set<const tir::BufferNode*> buffer_inlined;
       for (int i = 0; i < n_args; ++i) {
         tir::Var var = func->params[i];
         ObjectPath var_p = p->Attr("params")->ArrayIndex(i);
+        if (func->buffer_map.count(var)) {
+          tir::Buffer buffer = func->buffer_map[var];
+          ObjectPath buffer_p = p->Attr("buffer_map")->MapValue(var);
+          if (IsSimpleBuffer(buffer)) {
+            args.push_back(AssignDoc(DefineBuffer(buffer, *frame, d), NullOpt,
+                                     BufferAttn(buffer, buffer_p, *frame, d)));
+            buffer_inlined.insert(buffer.get());
+            continue;
+          }
+        }
```
It's likely that we will have to check the following conditions to make sure the syntactic sugars are correct (please add those corner cases into tests accordingly):
- The TIR var `func->params[i]` is used only once in this PrimFunc, i.e. in `func->params[i]`
- The buffer's data field `func->buffer_map[var]->data` is not shared with other buffers in `func->buffer_map`

I know the existing TVMScript printer isn't carefully implemented to take this into consideration, but let's do this right this time :-)

junrushao committed:
...

cyx-6 committed:
add check for shared buffer data

junrushao reviewed (approved):
LGTM!
